store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/addelements/handler/AddElementsHandler.java

```diff
@@ -0,0 +1,150 @@
+import uk.gov.gchq.gaffer.parquetstore.ParquetStore;
+import uk.gov.gchq.gaffer.parquetstore.operation.addelements.impl.AggregateAndSortTempData;
+import uk.gov.gchq.gaffer.parquetstore.operation.addelements.impl.GenerateIndices;
+import uk.gov.gchq.gaffer.parquetstore.operation.addelements.impl.WriteUnsortedData;
+import uk.gov.gchq.gaffer.spark.SparkUser;
+import uk.gov.gchq.gaffer.store.Context;
+import uk.gov.gchq.gaffer.store.Store;
+import uk.gov.gchq.gaffer.store.StoreException;
+import uk.gov.gchq.gaffer.store.operation.handler.OperationHandler;
+import uk.gov.gchq.gaffer.user.User;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+public class AddElementsHandler implements OperationHandler<AddElements> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AddElementsHandler.class);
+
+    @Override
+    public Void doOperation(final AddElements operation,
+                            final Context context, final Store store)
+            throws OperationException {
+        store.optimiseSchema();
+        final User user = context.getUser();
+        final SparkSession spark;
+        if (user instanceof SparkUser) {
+            spark = ((SparkUser) user).getSparkSession();
+        } else {
+            throw new OperationException("This operation requires the user to be of type SparkUser.");
+        }
+        configureSpark((ParquetStore) store, spark);
+        addElements(operation, (ParquetStore) store, spark);
+        return null;
+    }
+
+    private void configureSpark(final ParquetStore store, final SparkSession spark) {
+        final Integer numberOfOutputFiles = store.getProperties().getAddElementsOutputFilesPerGroup();
+        if (numberOfOutputFiles > Integer.parseInt(spark.conf().get("spark.sql.shuffle.partitions", "200"))) {
+            LOGGER.info("Setting the number of Spark shuffle partitions to " + numberOfOutputFiles);
+            spark.conf().set("spark.sql.shuffle.partitions", numberOfOutputFiles);
+        }
+
+        LOGGER.info("Setting the parquet file properties");
+        LOGGER.info("Row group size: " + store.getProperties().getRowGroupSize());
+        LOGGER.info("Page size: " + store.getProperties().getPageSize());
+        final Configuration hadoopConf = spark.sparkContext().hadoopConfiguration();
+        hadoopConf.setInt("parquet.block.size", store.getProperties().getRowGroupSize());
+        hadoopConf.setInt("parquet.page.size", store.getProperties().getPageSize());
+        hadoopConf.setInt("parquet.dictionary.page.size", store.getProperties().getPageSize());
+        hadoopConf.set("mapreduce.fileoutputcommitter.marksuccessfuljobs", "false");
+        hadoopConf.set("parquet.enable.summary-metadata", "false");
+    }
+
+    private void addElements(final AddElements addElementsOperation, final ParquetStore store, final SparkSession spark)
+            throws OperationException {
+        try {
+            final FileSystem fs = store.getFS();
```

Comment:
Where possible I reuse the FileSystem object, so this is only applicable in the two Callable classes where it is not getting the FileSystem from the Store.